File: src/49.group-anagrams.py
#
# @lc app=leetcode id=49 lang=python3
#
# [49] Group Anagrams
#
from typing import List
# @lc code=start
from collections import defaultdict
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        dic = defaultdict(list)

        for each in strs:
            key = self.sortCharacter(each)
            dic[key].append(each)
        
        return dic.values()

    def sortCharacter(self, word):
        return "".join(sorted(word))


# @lc code=end

# n: number of words
# k: maximum length of word in strs

# Time : O(N x KlogK), sorting -> logN
# Space : O(N x K)

File: Top Interview Questions/src/49.group-anagrams.js
/*
 * @lc app=leetcode id=49 lang=javascript
 *
 * [49] Group Anagrams
 */

// @lc code=start
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function(strs) {
  const anagrams = new Map();

  strs.forEach(each => {
    const key = makeSortedKey(each);
    if (anagrams.has(key)) {
      anagrams.get(key).push(each);
    } else {
      anagrams.set(key, [each]);
    }
  });

  return Array.from(anagrams.values());
};

/**
 * @param {string} word
 * @return {string}
 */
var makeSortedKey = (word) => {
  return word.split('').sort().join('');
};
// @lc code=end

